File: master-slave-apps/master-slave-app-js/app/controllers/item.js
const { createItem, getAllItems, getItemById, updateItem, deleteItem } = require('../models/item');

const createItemController = async (req, res) => {
  const { title, description, qty, price } = req.body;
  try {
    const item = await createItem(title, description, qty, price);
    res.status(200).json({
      code: 200,
      message: 'success',
      data: item
    });
  } catch (err) {
    res.status(500).json({
      code: 500,
      message: "Gagal membuat item",
      data: null
    });
  }
};

const getAllItemsController = async (req, res) => {
  try {
    const items = await getAllItems();
    if (!items) {
      return res.status(404).json({
      code: 404,
      message: 'Item tidak tersedia',
      data: items
    });
    }
    res.status(200).json({
      code: 200,
      message: 'success',
      data: items
    });
  } catch (err) {
    res.status(500).json({
      code: 500,
      message: 'Gagal mengambil data',
      data: null
    });
  }
};

const getItemByIdController = async (req, res) => {
  try {
    const item = await getItemById(req.params.id);

    if (!item) {
      return res.status(404).json({
        code: 404,
        message: 'Item tidak ditemukan',
        data: null
      });
    }
    res.status(200).json({
      code: 200,
      message: 'success',
      data: item
    });

  } catch (err) {
    res.status(500).json({
      code: 500,
      message: 'Gagal mengambil data',
      data: null
    });
  }
};

const updateItemController = async (req, res) => {
  const { title, description, qty, price } = req.body;
  try {
    const item = await updateItem(req.params.id, title, description, qty, price);
    if (!item) {
      return res.status(404).json({
        code: 404,
        message: 'Item tidak ditemukan',
        data: null
      });
    }
    
    res.status(200).json({
      code: 200,
      message: 'Item berhasil di update',
      data: item
    });
    
  } catch (err) {
    res.status(500).json({
      code: 500,
      message: 'Gagal update item',
      data: null
    });
  }
};

const deleteItemController = async (req, res) => {
  try {
    console.log(`Menerima permintaan DELETE untuk ID: ${req.params.id}`);
    const success = await deleteItem(req.params.id);
    if (!success) {
      return res.status(404).json({
        code: 404,
        message: 'Item tidak ditemukan',
        data: null
      });
    }
    res.status(200).json({
      code: 200,
      message: 'Item berhasil dihapus',
      data: null
    });
  } catch (err) {
    console.error(`Error saat DELETE ID: ${req.params.id}: ${err.message}`);
    res.status(500).json({
      code: 500,
      message: 'Gagal menghapus item',
      data: null
    });
  }
};

module.exports = {
  createItemController,
  getAllItemsController,
  getItemByIdController,
  updateItemController,
  deleteItemController
};

File: master-slave-apps/master-slave-app-go/app/handlers/item.go
package handlers

import (

	"database/sql"

	"apps/repositories"
	"apps/models"

)

type ItemHandler struct {
	repo *repositories.ItemRepository

}


func NewItemHandler(repo *repositories.ItemRepository) *ItemHandler {
	return &ItemHandler{repo:repo}
}


func (i *ItemHandler)FindByID(id string) models.Response {
	var (
		rsp = &models.Response{}
	)

	item, err := i.repo.FindByID(id)

	if err == sql.ErrNoRows {

		rsp.WithCode(404)
		rsp.WithMessage(`Item tidak ditemukan`)

		return *rsp
	}
	if err != nil {
		rsp.WithCode(500)
		rsp.WithMessage(`Gagal mengambil data`)

		return *rsp
	}
	rsp.WithCode(200)
	rsp.WithMessage(`success`)
	rsp.WithData(item)

	return *rsp
}


func (i *ItemHandler)FindAll() models.Response {
	var (
		rsp = &models.Response{}
	)

	items, err := i.repo.FindAll()

	if err == sql.ErrNoRows  || items == nil{

		rsp.WithCode(404)
		rsp.WithMessage(`Item tidak tersedia`)

		return *rsp
	}
	if err != nil {
		rsp.WithCode(500)
		rsp.WithMessage(`Gagal mengambil data`)

		return *rsp
	}
	rsp.WithCode(200)
	rsp.WithMessage(`success`)
	rsp.WithData(items)

	return *rsp
}

func (i *ItemHandler)Update(id string, item models.Item) models.Response {
	var (
		rsp = &models.Response{}
	)

	rowsAffected, err := i.repo.Update(id, &item)
	if err != nil {
		rsp.WithCode(500)
		rsp.WithMessage(`Gagal update item`)

		return *rsp
	}

	_ = rowsAffected

	itemU, err := i.repo.FindByID(id)
	if err == sql.ErrNoRows {

		rsp.WithCode(404)
		rsp.WithMessage(`Item tidak ditemukan`)

		return *rsp
	}
	if err != nil {
		rsp.WithCode(500)
		rsp.WithMessage(`Gagal update item`)

		return *rsp
	}


	itemRes := models.Item{
		ID: itemU.ID,
		Title: item.Title,
		Description: item.Description,
		Quantity: item.Quantity,
		Price: item.Price,
		CreatedAt: itemU.CreatedAt,
		UpdatedAt: itemU.UpdatedAt,
	}



	rsp.WithCode(200)
	rsp.WithMessage(`Item berhasil di update`)
	rsp.WithData(itemRes)


	return *rsp
}

func (i *ItemHandler)Store(item models.Item) models.Response {
	var (
		rsp = &models.Response{}
	)


	err := i.repo.Create(&item)
	
	
	if err != nil {
		rsp.WithCode(500)
		rsp.WithMessage(`Gagal membuat item`)

		return *rsp
	}


	rsp.WithCode(200)
	rsp.WithMessage(`success`)
	rsp.WithData(item)


	return *rsp
}



func (i *ItemHandler)Delete(id string) models.Response {
	var (
		rsp = &models.Response{}
	)

	rowsAffected, err := i.repo.Delete(id)
	if err != nil {
		rsp.WithCode(500)
		rsp.WithMessage(`Gagal menghapus item`)

		return *rsp
	}

	if rowsAffected == 0 {
		rsp.WithCode(404)
		rsp.WithMessage(`Item tidak ditemukan`)

		return *rsp
	}

	rsp.WithCode(200)
	rsp.WithMessage(`Item berhasil dihapus`)


	return *rsp
}